### ISSUE TITLE
Multiprocessing usage for hyperoptimisation

### ISSUE THREAD

**Issue by Axel-CH** [tensorflow/agents#437]
Hello,

I was using tf-agent succesfully with hyperopt in order to perform hyperparameter optimisation, with python 2.7.
To be able to use the latest features, I've decided to upgrade to python 3 and use the nightly version of tf-agent.

But now when I start my hyperopt worker I get this eror;

> RuntimeError: Unable to load multiprocessing context.
> 
> Please ensure that you properly initialize your program by wrapping your main()
> call:
> 
> def main(argv):
>   ...
> 
> if __name__ == '__main__':
>   tf_agents.system.multiprocessing.handle_main(main, extra_state_savers=...)
> 
> or, if using absl.app:
> 
> if __name__ == '__main__':
>   tf_agents.system.multiprocessing.handle_main(
>       functools.partial(absl.app.run, main), extra_state_savers=...)
> 
> 
> For unit tests, this also means wrapping your test.main using handle_test_main:
> 
> if __name__ == '__main__':
>   tf_agents.system.multiprocessing.handle_test_main(
>       tf.test.main, extra_state_savers=...)
> 
> or
> 
> if __name__ == '__main__':
>   tf_agents.system.multiprocessing.handle_test_main(
>       tf_agents.utils.test_utils.main, extra_state_savers=...)
> 
> If you are in interactive mode (e.g. python console, ipython, jupyter notebook)
> use:
> 
> tf_agents.system.multiprocessing.enable_interactive_mode(
>     extra_state_savers=...)
> 
> For more details on state savers, see the docstrings for
> `tf_agents.multiprocessing.handle_*` and:
> 
> https://pythonspeed.com/articles/python-multiprocessing/
> 
>   In call to configurable 'ParallelPyEnvironment' (<class 'tf_agents.environments.parallel_py_environment.ParallelPyEnvironment'>)
> ^CINFO:hyperopt.mongoexp:Caught signal 2, shutting down.
> Error in atexit._run_exitfuncs:
> Traceback (most recent call last):
>   File "/home/axel/anaconda3/envs/ag3/lib/python3.7/site-packages/ray/node.py", line 779, in _kill_process_type
> INFO:hyperopt.mongoexp:Caught signal 2, shutting down.

**Since my train_eval function is a python module directly called by hyperopt like this**:

    best = fmin(train_eval,
        space=space,
        algo=tpe.suggest,
        max_evals=FLAGS.max_evals,
        max_queue_len=1,
        trials=trials)

**I did not figured out yet how to properly initialize my multiprocessing context**. I think each hyperopt worker must have his own context.

The question is: where should I put something like this ?:
tf_agents.system.multiprocessing.handle_main(train_eval, `extra_state_savers=...)`

Thank for the help

**Comment by wookayin** [comment]
I think this is already described well in the documentation: https://github.com/tensorflow/agents/blob/v0.6.0/tf_agents/system/system_multiprocessing.py#L42-L74


```
tf_agents.system.multiprocessing.handle_main(train_eval, extra_state_savers=...)
```

Looks fine to me. Or alternatively you can have `
tf_agents.system.multiprocessing.enable_interactive_mode()` inside your main function.

**Comment by ebrevdo** [comment]
@Axel-CH did that work for you?

**Comment by Axel-CH** [comment]
@ebrevdo I did tried to add a wrapper around the "train_eval" function but it didn't work, maybe I did a mistake in my approach Bellow more details:

wrapper code :

```
def train_eval_wrapper(space,
        save_models=False,
        generate_plots=False,
        generate_dataframes=False,
        trial_desc="trial_description"):
    """Train eval wrapper to initialize multiprocessing context"""

    return multiprocessing.handle_main(partial(train_eval(space)))

```

hyperopt code calling "train_eval" (used as an objective function):

```
    best = fmin(train_eval_wrapper,
        space=space,
        algo=tpe.suggest,
        max_evals=FLAGS.max_evals,
        max_queue_len=1,
        trials=trials)
```

**Comment by wookayin** [comment]
Try using `tf_agents.system.multiprocessing.enable_interactive_mode()` only once (or applying the wrapper for the actual `__main__` function) before you start running hypot.

**Comment by Axel-CH** [comment]
I tried tf_agents.system.multiprocessing.enable_interactive_mode() in the main function, before the hyperoptimisation. I'm able to start the hypersearch.
But now I have a new issue, as I'm using the ray python library for multiprocessing usage. The interactive mode of tf_agent seem to mess with it.
The result is that after each hypersearch iteration I end up with a ghost process from ray consuming 90MiB of RAM.
So, after running a few hundred hypersearch iterations I'm consuming more that 200gb of ram.

I will try again with tf_agents.system.multiprocessing.handle_main(train_eval, extra_state_savers=...)

**Comment by Axel-CH** [comment]
I finally chose to disable ray for multiprocessing, my loading time during env initialisation is a little bit slower but I'm not having a memory leak anymore. Thanks for the help

**Comment by ebrevdo** [comment]
Yes; sorry about the issues w/ Ray multiprocessing.  We don't use Ray and would have a hard time debugging that.  In general, at least in Python having two libraries try to use multiprocessing is tough.  If your env is using multiprocessing internally, it may be easier to use a `BatchedPyEnvironment` instead of `Parallel`.  Batched uses multithreading and so if each sub-env is CPU-bound, then it's slower than Multiprocessing (due to the GIL).  But if each sub-env is just doing I/O (e.g., pipes via multiprocessing) then it may be less painful to use and just as fast or faster.

Closing this out for now, please reopen if you have additional issues.
